File: 2. Longest Substring Without Repeating Characters/solution.cpp
#include <string>
#include <vector>
#include <iostream>

class Solution {
public:
    int lengthOfLongestSubstring(std::string s) {
        if (s.empty()) return 0;
        
        // Use vector instead of unordered_map for better performance
        std::vector<int> lastSeen(128, -1);
        
        int maxLength = 0;
        int start = 0;
        
        for (int end = 0; end < s.length(); end++) {
            // If character was seen and its last position is >= start
            if (lastSeen[s[end]] >= start) {
                start = lastSeen[s[end]] + 1;
            }
            lastSeen[s[end]] = end;
            maxLength = std::max(maxLength, end - start + 1);
        }
        
        return maxLength;
    }
};

// Example usage
int main() {
    Solution solution;
    
    // Test cases
    std::cout << solution.lengthOfLongestSubstring("abcabcbb") << std::endl; // Output: 3
    std::cout << solution.lengthOfLongestSubstring("bbbbb") << std::endl;    // Output: 1
    std::cout << solution.lengthOfLongestSubstring("pwwkew") << std::endl;   // Output: 3
    std::cout << solution.lengthOfLongestSubstring("") << std::endl;         // Output: 0
    
    return 0;
}

File: 2. Longest Substring Without Repeating Characters/solution.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        if (s == null || s.length() == 0) return 0;
        
        // Use array instead of HashMap for better performance (assuming ASCII)
        int[] lastSeen = new int[128];
        // Initialize with -1 to handle first occurrence
        for (int i = 0; i < 128; i++) {
            lastSeen[i] = -1;
        }
        
        int maxLength = 0;
        int start = 0;
        
        for (int end = 0; end < s.length(); end++) {
            char currentChar = s.charAt(end);
            // If character was seen and its last position is >= start
            if (lastSeen[currentChar] >= start) {
                start = lastSeen[currentChar] + 1;
            }
            lastSeen[currentChar] = end;
            maxLength = Math.max(maxLength, end - start + 1);
        }
        
        return maxLength;
    }

    // Example usage
    public static void main(String[] args) {
        Solution solution = new Solution();
        
        // Test cases
        System.out.println(solution.lengthOfLongestSubstring("abcabcbb")); // Output: 3
        System.out.println(solution.lengthOfLongestSubstring("bbbbb"));    // Output: 1
        System.out.println(solution.lengthOfLongestSubstring("pwwkew"));   // Output: 3
        System.out.println(solution.lengthOfLongestSubstring(""));         // Output: 0
    }
}
